legacy/firmware/config.c

```diff
@@ -406,27 +407,36 @@ void config_init(void) {
 }
 
 void session_clear(bool lock) {
-  sessionSeedCached = secfalse;
-  memzero(&sessionSeed, sizeof(sessionSeed));
-  sessionIdCached = secfalse;
-  memzero(&sessionId, sizeof(sessionId));
+  for (uint8_t i = 0; i < MAX_SESSIONS_COUNT; i++) {
+    session_clearCache(sessionsCache + i);
+  }
+  activeSessionCache = NULL;
   if (lock) {
-    storage_lock();
+    session_lock();
   }
 }
 
+void session_clearCache(Session *session) {
+  session->usage = 0;
+  memzero(session->id, sizeof(session->id));
+  memzero(session->seed, sizeof(session->seed));
+  session->seedCached = false;
+}
+
+void session_lock(void) { storage_lock(); }
+
 static void get_u2froot_callback(uint32_t iter, uint32_t total) {
   layoutProgress(_("Updating"), 1000 * iter / total);
 }
 
 static void config_compute_u2froot(const char *mnemonic,
                                    StorageHDNode *u2froot) {
   static CONFIDENTIAL HDNode node;
+  static CONFIDENTIAL uint8_t seed[64];
   char oldTiny = usbTiny(1);
-  mnemonic_to_seed(mnemonic, "", sessionSeed,
-                   get_u2froot_callback);  // BIP-0039
+  mnemonic_to_seed(mnemonic, "", seed, get_u2froot_callback);  // BIP-0039
```

Comment:
@andrewkozlik I was a bit confused here. You have used sessionSeed here and called `session_clear` at the end, effectively destroying the current session. It seems alright, since it seems to be called only in load/reset/recovery/upgrade scanarios. Was it intentional? Why not simply declare a new array as I did?

Reply:
This doesn't seem familiar at all and I can't see what you are talking about, because the force-push destroyed the references.

Reply:
I am basically talking about our current master, see https://github.com/trezor/trezor-firmware/blob/2e2f5fd421ea5db06604ea09c320bdc9028961bd/legacy/firmware/config.c#L426 .

Reply:
```
045ef22d (Saleem Rashid   2017-12-18 21:16:05 +0000 270) static void storage_compute_u2froot(const char* mnemonic, StorageHDNode *u2froot) {
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 271)        static CONFIDENTIAL HDNode node;
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 272)        char oldTiny = usbTiny(1);
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 273)        mnemonic_to_seed(mnemonic, "", sessionSeed, get_u2froot_callback); // BIP-0039
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 274)        usbTiny(oldTiny);
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 275)        hdnode_from_seed(sessionSeed, 64, NIST256P1_NAME, &node);
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 276)        hdnode_private_ckd(&node, U2F_KEY_PATH);
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 277)        u2froot->depth = node.depth;
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 278)        u2froot->child_num = U2F_KEY_PATH;
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 279)        u2froot->chain_code.size = sizeof(node.chain_code);
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 280)        memcpy(u2froot->chain_code.bytes, node.chain_code, sizeof(node.chain_code));
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 281)        u2froot->has_private_key = true;
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 282)        u2froot->private_key.size = sizeof(node.private_key);
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 283)        memcpy(u2froot->private_key.bytes, node.private_key, sizeof(node.private_key));
bd660655 (Pavol Rusnak    2018-01-18 15:21:48 +0100 284)        memzero(&node, sizeof(node));
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 285)        session_clear(false); // invalidate seed cache
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 286) }
```

Reply:
Ahh, I did not dig deep enough. @prusnak do you remember anything about this?